homeassistant/components/tplink/config_flow.py

```diff
@@ -40,27 +62,102 @@ async def async_step_integration_discovery(
     ) -> FlowResult:
         """Handle integration discovery."""
         return await self._async_handle_discovery(
-            discovery_info[CONF_HOST], discovery_info[CONF_MAC]
+            discovery_info[CONF_HOST],
+            discovery_info[CONF_MAC],
+            discovery_info[CONF_DEVICE_CONFIG],
         )
 
-    async def _async_handle_discovery(self, host: str, mac: str) -> FlowResult:
+    @callback
+    def _update_config_if_entry_in_setup_retry(self, config: dict) -> None:
+        for entry in self._async_current_entries(include_ignore=True):
+            if entry.unique_id != self.unique_id:
+                continue
+            if entry.state == ConfigEntryState.SETUP_RETRY:
```

Comment:
This branch is a bit long, it would be cleaner to reverse the conditional here and continue